src/functions.jl

```diff
@@ -442,70 +460,79 @@ for T in (:Taylor1, :TaylorN)
-                mul!(c[k], -x, s[k-1])
-            end
-            @inbounds for i = 2:k
-                x = i * a[i]
+            @inbounds for i = 1:k
                 if $T == Taylor1
+                    x = i * a[i]
                     s[k] += x * c[k-i]
                     c[k] -= x * s[k-i]
                 else
-                    mul!(s[k], x, c[k-i])
-                    mul!(c[k], -x, s[k-i])
+                    mul!(s[k], a[i], c[k-i], scalar=i)
+                    mul!(c[k], a[i], s[k-i], scalar=-i)
                 end
             end
 
-            @inbounds s[k] = s[k] / k
-            @inbounds c[k] = c[k] / k
+            if $T == Taylor1
+                s[k] = s[k] / k
+                c[k] = c[k] / k
+            else
+                @inbounds div!(s[k], s[k], k)
+                @inbounds div!(c[k], c[k], k)
+            end
             return nothing
         end
 
         @inline function sincospi!(s::$T{T}, c::$T{T}, a::$T{T}, k::Int) where {T<:Number}
             if k == 0
                 a0 = constant_term(a)
-                @inbounds s[0], c[0] = sincospi( a0 )
+                if $T == Taylor1
+                    @inbounds s[0], c[0] = sincospi( a0 )
+                else
+                    @inbounds s[0][1], c[0][1] = sincospi( a0 )
+                end
                 return nothing
             end
-            aa = $T(pi * a[0], a.order)
-            @inbounds for ordQ in eachindex(a)
-                aa[ordQ] = pi * a[ordQ]
-            end
-            sincos!(s, c, aa, k)
+            mul!(a, pi, a, k)
+            sincos!(s, c, a, k)
             return nothing
         end
 
         @inline function tan!(c::$T{T}, a::$T{T}, c2::$T{T}, k::Int) where {T<:Number}
             if k == 0
                 @inbounds aux = tan( constant_term(a) )
-                @inbounds c[0] = aux
-                @inbounds c2[0] = aux^2
+                if $T == Taylor1
+                    @inbounds c[0] = aux
+                    @inbounds c2[0] = aux^2
+                else
+                    @inbounds c[0][1] = aux
+                    @inbounds c2[0][1] = aux^2
+                end
                 return nothing
             end
             zero!(c, k)
-            if $T == Taylor1
-                @inbounds c[k] = k * a[k] * c2[0]
-            else
-                @inbounds mul!(c[k], k * a[k], c2[0])
-            end
-            @inbounds for i = 1:k-1
+            @inbounds for i = 0:k-1
                 if $T == Taylor1
                     c[k] += (k-i) * a[k-i] * c2[i]
                 else
-                    mul!(c[k], (k-i) * a[k-i], c2[i])
+                    mul!(c[k], a[k-i], c2[i], scalar=k-i)
                 end
             end
-            @inbounds c[k] = a[k] + c[k]/k
+            # c[k] <- c[k]/k
+            if $T == Taylor1
+                div!(c, c, k, k) # Taylor1
+            else
+                div!(c[k], c[k], k) # TaylorN
+            end
```

Comment:
Same comment about order of `if` and `for`

Reply:
Done 👍